gax/src/main/java/com/google/api/gax/batching/v2/FlowController.java

```diff
@@ -0,0 +1,122 @@
+import com.google.api.gax.batching.v2.FlowControlException.MaxOutstandingRequestBytesReachedException;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.Semaphore;
+import javax.annotation.Nullable;
+
+/** Provides flow control capability. */
+class FlowController {
+
+  @Nullable private final Semaphore outstandingElementCount;
+  @Nullable private final Semaphore64 outstandingByteCount;
+  private final int maxOutstandingElementCount;
+  private final long maxOutstandingRequestBytes;
+  private final boolean isBlocking;
+
+  FlowController(FlowControlSettings settings) {
+    isBlocking =
+        settings.getLimitExceededBehavior() == FlowControlSettings.LimitExceededBehavior.Block;
+    switch (settings.getLimitExceededBehavior()) {
+      case ThrowException:
+      case Block:
+        break;
+      case Ignore:
+        this.maxOutstandingElementCount = 0;
+        this.maxOutstandingRequestBytes = 0;
+        this.outstandingElementCount = null;
+        this.outstandingByteCount = null;
+        return;
+      default:
+        throw new IllegalArgumentException(
+            "Unknown LimitBehaviour: " + settings.getLimitExceededBehavior());
+    }
+
+    this.maxOutstandingElementCount = settings.getMaxOutstandingElementCount();
+    if (maxOutstandingElementCount == 0) {
+      outstandingElementCount = null;
+    } else {
+      outstandingElementCount = new Semaphore(maxOutstandingElementCount, true);
+    }
+
+    this.maxOutstandingRequestBytes = settings.getMaxOutstandingRequestBytes();
+    if (maxOutstandingRequestBytes == 0) {
+      outstandingByteCount = null;
+    } else if (isBlocking) {
+      outstandingByteCount = new BlockingSemaphore(maxOutstandingRequestBytes);
+    } else {
+      outstandingByteCount = new NonBlockingSemaphore(maxOutstandingRequestBytes);
+    }
+  }
+
+  public void reserve(long bytes) {
+    Preconditions.checkArgument(bytes >= 0);
+    try {
+      if (outstandingElementCount != null) {
+        if (isBlocking) {
+          outstandingElementCount.acquire();
+        } else if (!outstandingElementCount.tryAcquire()) {
+          throw new MaxOutstandingElementCountReachedException(maxOutstandingElementCount);
+        }
+      }
+
+      // Will always allow to send a request even if it is larger than the flow control limit,
+      // if it doesn't then it will deadlock the thread.
+      if (outstandingByteCount != null) {
+        long permitsToDraw = Math.min(bytes, maxOutstandingRequestBytes);
+        if (!outstandingByteCount.acquire(permitsToDraw)) {
+          if (outstandingElementCount != null) {
+            outstandingElementCount.release();
+          }
+          throw new MaxOutstandingRequestBytesReachedException(maxOutstandingRequestBytes);
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new MaxOutstandingElementCountReachedException(
```

Comment:
This looks weird, why is interruption bubbled up as a MaxOutstandingElementCountReachedException? Also, shouldn't the thread interrupt flag be reset?

Reply:
This exception is coming from `Semaphore#acquire` which is for elementCount. We were resetting the interrupted flag for `BlockingSemaphore#acquire`.

Also updated it to wrap this `InterruptedException` with `FlowControlException`.